### DISCUSSION TITLE
Why does Fastlane `sh` command add escape characters?

### DISCUSSION THREAD

**Discussion by simon-assistiq** [fastlane/fastlane#21676] [Q&A] [answered]
In the [`sh` documentation](https://docs.fastlane.tools/actions/sh/) there's this example

```
sh("git", "commit", "-m", "My message")
```

This is useful for splitting up a long command, especially if you want to split it up onto multiple lines. However, in certain cases when I've split up my command like this, the resulting command has spaces and special charaters needlessly escaped, breaking the command.

This works fine
```
sh('git log -2')
```

But if I do this:

```
sh('git log -2', "--pretty='format:%h'")
```

the resulting command is:

```
git\ log\ -2 --pretty\=\'format:\%h\'
```

i.e. `sh` suddenly decides to escape all the spaces, the `=`, the `%` and the `'`.

Is this by design? Is there a way to prevent it / am I doing something wrong here?

Found the same issue was discussed [here](https://github.com/fastlane-community/fastlane-plugin-s3/issues/77), though the posted solution doesn't attempt to explain why this happens.

I am using Fastlane `2.217.0`, on macOS `13.6.2 (22G320)` in the built-in zsh Terminal.

Thank you

**Comment by lacostej** [comment]
CC @AliSoftware :)

**Comment by AliSoftware** [comment]
This is by design indeed, because that form of `sh` where you pass multiple parameters (instead of a single long String) works the same as Ruby's build in Kernel#open and Popen3 and all those family of methods, in that in this form the first parameter is the name of the command, and the remaining parameters are the arguments of the command, one argument per parameter.

So `sh("cat", "file.txt", "hello world.txt")` will call the command `cat` with 2 arguments, the first one being `file.txt` and the second being `hello world.txt`.

This form of `sh` does _not_ pass any of those strings thru a shell interpreter. This means that you don't have to shell-escape the parameters to prevent the shell from interpreting a space or special character (like you'd need to do if you used the form that only takes a single string in parameter). But that also means that all characters in this form are not interpreted and taken as literals in the command and its arguments. Especially, a space in any of those arguments is just seen as a literal space character in the argument passed to the `cat` command, not as a separator between arguments (that would require escaping if you wanted it to be a literal space instead).

Maybe some examples will help clarify that a bit better:

```ruby
sh('cat', 'foo', 'hello world')
# command is: cat
# arg 1 passed to cat: foo
# arg 2 passed to cat: hello world
```

```ruby
sh('cat foo hello world')
# single string, so it's passed to a shell interpreter first to split it into command + arguments, resulting in:
# command: cat
# arg 1: foo
# arg 2: hello
# arg 3: world
```

```ruby
sh('cat foo hello\ world')
# single string, so it's passed to a shell interpreter first to split it into command + arguments based on the spaces and escape sequences, resulting in:
# command: cat
# arg 1: foo
# arg 2: hello world
```

```ruby
sh("git log -2 --pretty='format:%h")
# Single string, so it's passed to a shell interpreter to split into command + arguments, resulting in:
# command: git
# arg 1: log
# arg 2: -2
# arg 3: --pretty='format:%h'
```

```ruby
sh('git log -2', "--pretty='format:%h")
# command/program to invoke: git log -2
# arg 1: --pretty='format:%h'
```
In 👆 this one, it will try to invoke a command whose name is "git log -2". There is no such command in $PATH, as the command program you are looking for is `git`. For comparison, it's a bit like if you were telling it to open an app called `git log -2.app` instead of telling to open an app called `git.app` and pass it arguments `log` and `-2`.

This is also why `sh` adds those backslash escapes here. Because in this form of `sh`, the first argument of `sh` is supposed to be the name of the command/program you want to invoke. And if you indeed wanted to invoke a program whose full name was `git log -2` then you'd need to escape the spaces in the name of the program when typing it in a shell.

---

If you are using this form of `sh` what you really want is only pass the program name as first parameter, then pass each argument individually, as individual parameters.

```ruby
# For example, this won't work either:
sh('git', 'log -2', '--pretty=format:%h')
# because it will now invoke the right program (git), it will now pass 'log -2' as a single argument to the command, not as two separate arguments.
# So this invocation of sh would be equivalent if you typing this in the terminal:
git log\ -2 --pretty:format=%h
# Or equivalently:
git "log -2" "--pretty:format=%h"
```
Instead what you want to use is this:
```
sh('git', 'log', '-2', '--pretty=format:%h')
```
Thus passing each argument you want to pass to the `git` program as a separate parameter to `sh`.

_Again, all of this is the same behavior as build-in Ruby methods that similarly deal with opening programs, like `Kernel#open`, `system(…)`, `Popen3#open3` and all those. So you can find similar information on those methods in Ruby's own documentation if you want to learn more on that syntax._

**Reply by AliSoftware** [reply]
Also side note that I'd personally recommend using the form where you pass arguments individually, rather that passing a single long string, because:

 - It avoids the needless use of a shell interpreter to split your long single String into a command and its arguments (which involves the shell interpreter parsing the String, detecting spaces as argument separators, considering escape sequences and quotes in the string…), while usually as a caller you already know how the individual arguments so you might as well pass them individually
 - This avoids having to remember to shell_escape each argument that could potentially contain a special character that could be interpreted by the shell interpreter (like a space or quote) if you were to use a single string
 - And thus it avoids the risk of security issues with malicious parameter injection via user input.

For example:

```ruby
workspace = … # name of a xcworkspace file provided by the user
sh("xcodebuild -workspace #{workspace}")
# -> this form of sh will use a shell interpreter but the `workspace` variable is not escaped, so if it contains spaces or special characters this will be interpreted by the shell and the command will not work as expected

# Besides, that allows a malicious user to pass `workspace = "Foo.xcworkspace; rm -rf $HOME"` and since that will not be escaped and instead evaluated by the shell interpreter as is, could lead to data loss and security issue.

#---

sh("xcodebuild -workspace '#{workspace}'")
# -> Maybe you thought putting single quote around the variable would help for if the workspace name contained spaces. But then it will
 fail if the workspace file contains quotes in its name now

#---
sh("xcodebuild -workspace #{workspace.shellescape}")
# -> This is a bit better security-wise because at lease you properly escape the string representing the workspace name before passing it to the shell interpreter which will unescape it. So even if the `workspace` variable contained spaces, those will be escaped before passing them to the shell which will then interpret them as literal space characters and not argument separators.

# But then the commands can become quite long to type in Ruby if you have to `#{variable.shellescape}` every variable you want to inject in your command so that the shell interpreter unescape it… and having that whole dance of escaping so that shell unescape it is a bit of a waste if you already know how the arguments should be split and passed to the command individually; why rely on a shell interpreter to do that split of arguments if you already know?

# So it's much simpler and safer to just bypass the shell interpreter (and its potential security issues if you fail to escape the inputs) in the first place, by using the second form of sh instead directly:

sh('xcodeproj', '-workspace', workspace)

# -> No more need to escape the workspace variable, will still work even if that variable contain special characters like spaces and quote that a shell could have interpreted while we wouldn't, no security risk of parameter injection
```

**Reply by simon-assistiq** [reply]
Thanks for a detailed response!

When I run this:
```
sh('git', 'log', '-2', '--pretty=format:%h')
```

The resulting command in the Terminal looks like:

```
git log -2 --pretty\=format:\%h
```

so there are still some extra escapes there. It happens to work in this case though. But I also have a `--grep='DEV'` in my case…

Fastfile:
```
sh('git', 'log', '-2', "--grep='DEV'", '--pretty=format:%h')
```

Resulting command:
```
git log -2 --grep\=\'DEV\' --pretty\=format:\%h
```

and the output is empty. Whereas if I try the command above without the `\` escapes it does return the two commits I want.

**Reply by AliSoftware** [reply, marked as answer]
Interesting case!

I believe this is also expected. The fact that running the command in a shell / Terminal with `--grep='DEV'` works is because the single quotes around `DEV` are also interpreted by the shell. So in the end the shell interpreter in your Terminal would remove interpret those single quotes as just quotes that ensure that `DEV` is a single word and that any special character within those quotes are not interpreted by the shell itself, and it will end up passing `--grep=DEV` as the argument to the `git` command.

But when you're passing this as explicit parameters to `sh` (or `system` or `Kernel#open`, etc), since in that form there is no shell interpreter getting in the way processing the string, then you should not include the single-quotes around the term you want to grep.

You can confirm this by checking that this behavior is the same when you use Ruby's build-in `system` command in a Ruby interpreter like `irb` or `pry` for example—to confirm that this is not specific to fastlane and its `sh` helper method:

```ruby
system('git', 'log', '-2', "--grep='DEV'", '--pretty=format:%h')
# --> Will fail to find commits as it's searching for `'DEV'` literal, including seeing the single quotes as literal characters to search for

system('git', 'log', '-2', "--grep=DEV", '--pretty=format:%h')
# --> Will find commits containing the word `DEV`, without including quotes in that literal that is being searched
```

So because in a shell (like when you type the command in a Terminal), `'` single quote characters are interpreted by the shell (to mean "consider the quoted string as being a single string part of the same argument, without interpreting special characters or spaces within the quotes but instead considering them as literals"), that is why:
 - It works even when you use those single quotes while typing that command in a shell (like in your Terminal.app), or when you use those single quotes while calling `system` or `sh` with a single string (that is gonna be interpreted by the shell interpreter to be split into arguments)
 - But when you pass those as explicit parameters to `system` or `sh`, since that won't go thru a shell interpreter, those single quotes will just be seen as literal `'` characters (and you don't need to use those single quotes to protect against spaces in the search term when you pass that argument as an explicit parameter, since the shell interpreter won't be involved)
 - Which is also why it escapes those as `\'` when printing the command you'd type to do the equivalent thing if typed in a shell interpreter.

Because when calling `sh('git', 'log', '-2', "--grep='DEV'", '--pretty=format:%h')` you really mean to say "call the command named `git`, and pass it the `--grep='DEV'` **literal** string as 3rd parameter". And since that's what you ask to `sh` by using such a function call with explicit arguments, you are indeed telling it to use `'DEV'`—with the single quotes seen as literal characters and all—as the search term. And indeed, the equivalent to do the exact same if you were to type that command in a shell (like in Terminal.app) would be to use `--grep=\'DEV\'`, escaping those single quotes so that the shell interpreter would see those as literal `'` apostrophe characters instead of seeing them as quotes wrapping some string.

Since `sh('git', 'log', '-2', "--grep='DEV'", '--pretty=format:%h')` is asking to search for `'DEV'` with literal single quotes included, then Ruby is correct to say that the equivalent shell command would be `git log -2 --grep=\'DEV\' --pretty=format:\%h`.

---

As to why the shell interpreter also escapes `\=` and `\%` as well… I don't know all the rules and list of which characters are considered "potential special characters that needs escaping" in shell interpreters, but according to that it seems that the standard considers `=` and `%` to be special characters that needs escaping, and that's why it adds backslash before them.

Obviously, you can see that if you try those commands in your shell without escaping `=` nor `%` they will still work without issues. But my guess is that those characters **_might_** be interpreted as special characters with special meaning even if only in some contexts but not all (for example, `=` is used for declaring variables in shell as in `git=foo` would declare a variable `git`, or `%` can be used in specific contexts like to reference bash job numbers or doing bash arithmetics). And since it doesn't hurt to escape them systematically (as the shell interpreter will replace `\=` with a literal `=` in all cases), that's what [Ruby's shellescape method seems to have decided to do](https://docs.ruby-lang.org/en/master/Shellwords.html):

```bash
$ irb
irb(main):001:0> require 'shellwords'
=> true
irb(main):002:0> puts "--grep='DEV'".shellescape
--grep\=\'DEV\'
=> nil
irb(main):003:0> puts '--pretty=format:%h'.shellescape
--pretty\=format:\%h
=> nil
```

**Reply by simon-assistiq** [reply]
Ah, I see, so the quotation marks aren't needed…

Even if I insert a variable  (e.g. `"--grep=#{some_variable}"`) which contains spaces they will automatically be escaped.

And as you mention the escaping of the `=` and `%` doesn't cause any problems.

Thank you again

**Reply by AliSoftware** [reply]
> Ah, I see, so the quotation marks aren't needed…
> 
> Even if I insert a variable with spaces they will automatically be escaped.

Yep exactly.
